internal/domain/power-status.go

```diff
@@ -18,16 +19,707 @@
  * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  * OTHER DEALINGS IN THE SOFTWARE.
+ *
  */
 
 package domain
 
 import (
-	"github.com/Cray-HPE/hms-power-control/internal/model"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/Cray-HPE/hms-power-control/internal/credstore"
+	"github.com/Cray-HPE/hms-power-control/internal/storage"
+	pcshsm "github.com/Cray-HPE/hms-power-control/internal/hsm"
+	pcsmodel "github.com/Cray-HPE/hms-power-control/internal/model"
+	rf "github.com/Cray-HPE/hms-smd/pkg/redfish"
+	trsapi "github.com/Cray-HPE/hms-trs-app-api/pkg/trs_http_api"
+	"github.com/Cray-HPE/hms-xname/xnametypes"
+
+	"github.com/sirupsen/logrus"
+)
+
+
+// Power monitoring framework.  The general flow of this framework:
+//
+// o Read all components and supporting info from HSM (includes various RF
+//   URLs), and store them in an internal memory-cached map.
+// o Periodically use this map to figure out which BMCs to query for HW state.
+// o Query BMCs for HW state, and store the results both in the in-memory
+//   cached component map and also in ETCD storage.
+// o If user asks for power status on a set of components, it is fetched from
+//   the ETCD-stored data.  This makes it possible for any instance of a 
+//   multi-instance PCS service to be able to serve data retrieved by any other
+//   instance.
+
+
+// This is what gets stored in the in-memory component map.
+
+type componentPowerInfo struct {
+	PSComp      pcsmodel.PowerStatusComponent
+	HSMData     pcshsm.HsmData
+	BmcUsername string
+	BmcPassword string
+}
+
+const (
+	PSTATE_KEYRANGE_START = "x0"
+	PSTATE_KEYRANGE_END   = "xz"
 )
 
-func GetPowerStatus(xnames []string, psf model.PowerStateFilter, msf model.ManagementStateFilter) (pb model.Passback) {
-	//TODO stuff here!
-	pb = model.BuildSuccessPassback(501, "GetPowerStatus")
-	return pb
+var hwStateMap = make(map[string]*componentPowerInfo)
+var glogger *logrus.Logger
+var hsmHandle *pcshsm.HSMProvider
+var kvStore *storage.StorageProvider
+var ccStore *credstore.CredStoreProvider
+var distLocker *storage.DistributedLockProvider
+var tloc *trsapi.TrsAPI
+var pmSampleInterval time.Duration
+var distLockMaxTime time.Duration
+var pstateMonitorRunning bool
+var serviceRunning *bool
+var vaultEnabled = false
+var httpTimeout = 15 //seconds
+
+
+/////////////////////////////////////////////////////////////////////////////
+//                P U B L I C   F U N C T I O N S
+/////////////////////////////////////////////////////////////////////////////
+
+
+// Initialize and start power state monitoring.
+//
+// domGlb:            Domain global object (holds lots of handles)
+// distLockMaxTimeIn: Max time to hold dist'd lock before forfeiting
+// loggerIn:          Logrus logger object.  Will create if nil.
+// sampleInterval:    Time between HW status sampling.
+// Return:            nil on success, else error message.
+
+func PowerStatusMonitorInit(domGlb *DOMAIN_GLOBALS,
```

Comment:
Are you holding off on adding this to main() for a future integration mod?

Reply:
It's coming in my next push.   I'm testing it right now.